README.rst
===============================
Xapo SDK & Tools
===============================

.. image:: https://badge.fury.io/py/xapo_sdk.png
    :target: http://badge.fury.io/py/xapo_sdk

.. image:: https://travis-ci.org/xapo/python-sdk.svg
        :target: https://travis-ci.org/xapo/python-sdk

.. image:: https://pypip.in/d/xapo_sdk/badge.png
        :target: https://pypi.python.org/pypi/xapo_sdk


Xapo's bitcoin sdk & tools

This is the Python version of the Xapo's SDK & Widget Tools. These tools allow you (Third Party Application, TPA) to easily develop your bitcoins applications to manage *Accounts, Wallets, Transactions*, etc., or embed tools like *Payments Buttons, Donation Buttons* and other kind of widgets into your web application using your language of choice. In this way, tedious details like encryption and HTML snippet generation are handled for you in a simple and transparent way.

* Free software: BSD license
* Documentation: http://xapo-sdk.readthedocs.org.


Features
--------

* *Iframe* and *Div* HTML widgets snippet generator.

HISTORY.rst
.. :changelog:

History
-------

0.2.2 (2014-11-14)
---------------------

* Added customization options and callbacks to Xapo micro payment tool.

0.2.1 (2014-11-04)
---------------------

* Added xapo_api module, including the credit api v0 (this is about to change soon in the incoming v1 api).
* Introduced xapo_tool.PaymentType constants.

0.2.0 (2014-11-01)
---------------------

* Having a TPA is not longer mandatory to provide a micro payments widget. Parameters app_id and app_secret are now optional. 

0.0.1 (2014-10-18)
---------------------

* First release on PyPI.
* Includes:
    * iframe widget
    * div widget

tests/test_xapo_api.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_xapo_api
----------------------------------

Tests for `xapo_api` module.
"""

import unittest2

import uuid
from xapo_sdk import xapo_api

TEST = False


# TODO set your credentials ans TEST to True
@unittest2.skipIf(not TEST, "Set your credentials and TEST to True")
class TestXapo_tools(unittest2.TestCase):
    def setUp(self):
        self.api = xapo_api.API('https://api.xapo.com/v1',
                                'your app id',
                                'your app secret')

    def test_credit(self):
        res = self.api.credit(to='sample@xapo.com', amount=1,
                              currency=xapo_api.Currency.SAT,
                              comments='Sample deposit',
                              request_id=uuid.uuid1().hex)

        print(res)

        self.assertTrue(res['success'])
        self.assertEquals('Success', res['code'])

    def test_credit_bad_amount(self):
        res = self.api.credit(to='sample@xapo.com', amount=-0.5,
                              comments='Sample deposit',
                              request_id=uuid.uuid1().hex)

        print(res)

        self.assertFalse(res['success'])
        self.assertEquals('InvalidAmount', res['code'])

    def test_credit_missing_to(self):
        res = self.api.credit(to='', amount=0.5,
                              comments='Sample deposit',
                              request_id=uuid.uuid1().hex)

        print(res)

        self.assertFalse(res['success'])
        self.assertEquals('Failure', res['code'])


tests/test_xapo_sdk.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_xapo_sdk
----------------------------------

Tests for `xapo_sdk` module.
"""

import unittest

from xapo_sdk import xapo_utils
from xapo_sdk import xapo_tools
import re


class TestXapo_sdk(unittest.TestCase):

    def setUp(self):
        self.mp = xapo_tools.MicroPayment(
            "http://dev.xapo.com:8089/pay_button/show",
            "b91014cc28c94841",
            "c533a6e606fb62ccb13e8baf8a95cbdc")
        self.mp_notpa = xapo_tools.MicroPayment(
            "http://dev.xapo.com:8089/pay_button/show")

    def test_encrypt(self):
        json = """{"sender_user_id":"s160901",\
"sender_user_email":"fernando.taboada@gmail.com",\
"sender_user_cellphone":"",\
"receiver_user_id":"r160901"\
"receiver_user_email":"fernando.taboada@xapo.com",\
"tip_object_id":"to160901",\
"amount_SAT":"",\
"timestamp":1410973639125}"""
        enc = xapo_utils.encrypt(json, "bc4e142dc053407b0028accffc289c18")
        expected = b'rjiFHpE3794U23dEKNyEz3ukF5rhVxtKzxEnZq8opuHoRH5eA/XOEbROE\
zf5AYmyQ5Yw6cQLSVMx/JgENrNKVK268n3o1kOIxEpupaha2wYXLqIqU8Ye7LFQz7N\
vQNPzfyOSPWnBQ/JUCSKsCiCz45VoK511B/RMz33mjJMF7s2qkQlCjukzZ1rry5gRF\
XaOxGgK23nnvlx7YX+3YVyZpLT+dpwy1x0gy86+Iwq/bjSh/V65R9Og71Zm47MjxKr\
scoJWV5+pzlSKKx2bTCqBX1sI+gedozC1GEdKINy9Ug67XvtqyXMjKBlxZIC4vTw9q\
gjV/sbkB7LZ2ShWFIBJRQ=='

        print(json)
        print(enc.decode())
        self.assertEquals(expected, enc)

    def test_build_iframe_widget(self):
        mpc = xapo_tools.MicroPaymentConfig(
            sender_user_email="sender@xapo.com",
            sender_user_cellphone="+5491112341234",
            receiver_user_id="r0210",
            receiver_user_email="fernando.taboada@xapo.com",
            pay_object_id="to0210",
            amount_BIT=0.01,
            pay_type="Tip")
        iframe = self.mp.build_iframe_widget(mpc)
        regex = r'\n<iframe(.*)button_request(.*)>(.*)</iframe>\n'

        print(iframe)
        self.assertNotEqual(None,
                            re.match(regex, iframe, re.MULTILINE | re.DOTALL))

    def test_build_iframe_widget_notpa(self):
        mpc = xapo_tools.MicroPaymentConfig(
            sender_user_email="sender@xapo.com",
            sender_user_cellphone="+5491112341234",
            receiver_user_id="r0210",
            receiver_user_email="fernando.taboada@xapo.com",
            pay_object_id="to0210",
            amount_BIT=0.01,
            pay_type="Tip")
        iframe = self.mp_notpa.build_iframe_widget(mpc)
        regex = r'\n<iframe(.*)payload(.*)>(.*)</iframe>\n'

        print(iframe)
        self.assertNotEqual(None,
                            re.match(regex, iframe, re.MULTILINE | re.DOTALL))

    def test_build_div_widget(self):
        mpc = xapo_tools.MicroPaymentConfig(
            sender_user_email="sender@xapo.com",
            sender_user_cellphone="+5491112341234",
            receiver_user_id="r0210",
            receiver_user_email="fernando.taboada@xapo.com",
            pay_object_id="to0210",
            amount_BIT=0.01,
            pay_type="Donate")
        div = self.mp.build_div_widget(mpc)
        regex = r"""
<div id="tipButtonDiv" class="tipButtonDiv"></div>
<div id="tipButtonPopup" class="tipButtonPopup"></div>
<script>(.*)button_request(.*)</script>
"""
        print(div)
        self.assertNotEqual(None, 
                            re.match(regex, div, re.MULTILINE | re.DOTALL))

    def test_build_div_widget_notpa(self):
        mpc = xapo_tools.MicroPaymentConfig(
            sender_user_email="sender@xapo.com",
            sender_user_cellphone="+5491112341234",
            receiver_user_id="r0210",
            receiver_user_email="fernando.taboada@xapo.com",
            pay_object_id="to0210",
            amount_BIT=0.01,
            pay_type="Donate")
        div = self.mp_notpa.build_div_widget(mpc)
        regex = r"""
<div id="tipButtonDiv" class="tipButtonDiv"></div>
<div id="tipButtonPopup" class="tipButtonPopup"></div>
<script>(.*)payload(.*)</script>
"""

        print(div)
        self.assertNotEqual(None, 
                            re.match(regex, div, re.MULTILINE | re.DOTALL))

    def tearDown(self):
        pass

if __name__ == '__main__':
    unittest.main()


tests/test_xapo_tools.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_xapo_tools
----------------------------------

Tests for `xapo_tools` module.
"""

import unittest2

from xapo_sdk import xapo_tools
from xapo_sdk.xapo_tools import PaymentType
import re

TEST = False


# TODO set your credentials ans TEST to True
@unittest2.skipIf(not TEST, "Set your credentials and TEST to True")
class TestXapo_tools(unittest2.TestCase):

    def setUp(self):
        self.mp = xapo_tools.MicroPayment(
            "https://mpayment.xapo.com/pay_button/show"
            "your app id",
            "your app secret")
        self.mp_notpa = xapo_tools.MicroPayment(
            "https://mpayment.xapo.com/pay_button/show")

    def test_build_iframe_widget(self):
        config = xapo_tools.MicroPaymentConfig(
            sender_user_email="sender@xapo.com",
            sender_user_cellphone="+5491112341234",
            receiver_user_id="r0210",
            receiver_user_email="fernando.taboada@xapo.com",
            pay_object_id="to0210",
            amount_BIT=1,
            pay_type=PaymentType.TIP,
            end_mpayment_uri="http://localhost:9000",
            redirect_uri="http://localhost:9000")
        customization = xapo_tools.MicroPaymentCustomization(
            login_cellphone_header_title="Test MicroPayment",
            predefined_pay_values="1,5,10",
            button_css="red")

        iframe = self.mp.build_iframe_widget(config, customization)
        regex = r'\n<iframe(.*)button_request(.*)>(.*)</iframe>\n'

        print(iframe)
        self.assertNotEqual(None,
                            re.match(regex, iframe, re.MULTILINE | re.DOTALL))

    def test_build_iframe_widget_notpa(self):
        config = xapo_tools.MicroPaymentConfig(
            sender_user_email="sender@xapo.com",
            sender_user_cellphone="+5491112341234",
            receiver_user_id="r0210",
            receiver_user_email="fernando.taboada@xapo.com",
            pay_object_id="to0210",
            amount_BIT=0.01,
            pay_type=PaymentType.TIP,
            end_mpayment_uri="http://localhost:9000",
            redirect_uri="http://localhost:9000")
        customization = xapo_tools.MicroPaymentCustomization(
            login_cellphone_header_title="Test MicroPayment",
            predefined_pay_values="1,5,10",
            button_css="grey")

        iframe = self.mp_notpa.build_iframe_widget(config, customization)
        regex = r'\n<iframe(.*)payload(.*)>(.*)</iframe>\n'

        print(iframe)
        self.assertNotEqual(None,
                            re.match(regex, iframe, re.MULTILINE | re.DOTALL))

    def test_build_div_widget(self):
        config = xapo_tools.MicroPaymentConfig(
            sender_user_email="sender@xapo.com",
            sender_user_cellphone="+5491112341234",
            receiver_user_id="r0210",
            receiver_user_email="fernando.taboada@xapo.com",
            pay_object_id="to0210",
            amount_BIT=0.01,
            pay_type=PaymentType.DONATE,
            end_mpayment_uri="http://localhost:9000",
            redirect_uri="http://localhost:9000")
        customization = xapo_tools.MicroPaymentCustomization(
            login_cellphone_header_title="Test MicroPayment",
            predefined_pay_values="1,5,10",
            button_css="red")

        div = self.mp.build_div_widget(config, customization)
        regex = r"""
<div id="tipButtonDiv" class="tipButtonDiv"></div>
<div id="tipButtonPopup" class="tipButtonPopup"></div>
<script>(.*)button_request(.*)</script>
"""
        print(div)
        self.assertNotEqual(None,
                            re.match(regex, div, re.MULTILINE | re.DOTALL))

    def test_build_div_widget_notpa(self):
        config = xapo_tools.MicroPaymentConfig(
            sender_user_email="sender@xapo.com",
            sender_user_cellphone="+5491112341234",
            receiver_user_id="r0210",
            receiver_user_email="fernando.taboada@xapo.com",
            pay_object_id="to0210",
            amount_BIT=0.01,
            pay_type=xapo_tools.PaymentType.DONATE,
            end_mpayment_uri="http://localhost:9000",
            redirect_uri="http://localhost:9000")
        customization = xapo_tools.MicroPaymentCustomization(
            login_cellphone_header_title="Test MicroPayment",
            predefined_pay_values="1,5,10",
            button_css="grey")

        div = self.mp_notpa.build_div_widget(config, customization)
        regex = r"""
<div id="tipButtonDiv" class="tipButtonDiv"></div>
<div id="tipButtonPopup" class="tipButtonPopup"></div>
<script>(.*)payload(.*)</script>
"""

        print(div)
        self.assertNotEqual(None,
                            re.match(regex, div, re.MULTILINE | re.DOTALL))

    def tearDown(self):
        pass

if __name__ == '__main__':
    unittest.main()


xapo_sdk/xapo_api.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from __future__ import absolute_import, division, print_function, \
    with_statement, unicode_literals

import requests
import time
import json

from xapo_sdk import xapo_utils


class Currency:
    BTC = "BTC"
    SAT = "SAT"


class API:
    __service_url = None
    __app_id = None
    __app_secret = None

    __credit_resource = "/credit/"

    def __init__(self, service_url, app_id, app_secret):
        self.__service_url = service_url
        self.__app_id = app_id
        self.__app_secret = app_secret

    def credit(self, to, amount, request_id, currency=Currency.BTC,
               comments="", subject=""):
        timestamp = int(round(time.time() * 1000))
        payload = {
            "to": to,
            "currency": currency,
            "amount": amount,
            "comments": comments,
            "subject": subject,
            "timestamp": timestamp,
            "unique_request_id": request_id
        }

        json_payload = json.dumps(payload)
        encrypted_payload = xapo_utils.encrypt(json_payload, self.__app_secret)
        query = {"appID": self.__app_id, "hash": encrypted_payload}

        r = requests.get(url=self.__service_url + self.__credit_resource,
                         params=query)

        return r.json()


xapo_sdk/xapo_tools.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from __future__ import absolute_import, division, print_function, \
    with_statement, unicode_literals

from future.standard_library import hooks
with hooks():
    from urllib.parse import urlencode

import time
import textwrap
import json
from xapo_sdk import xapo_utils


class PaymentType:
    NONE = ''
    TIP = 'Tip'
    DEPOSIT = 'Deposit'
    PAY = 'Pay'
    DONATE = 'Donate'
    OAUTH = 'Oauth'


class MicroPaymentConfig:
    """ Micro payment button configuration options.

    This class is intended to be a placeholder for micro payments
    buttons configuration but also serves for documenting. A dictionary
    with the intended fields would give the same results.

    Attributes:
        sender_user_id (str): The id of the user sending the payment.

        sender_user_email (str, optional): The email of the user sending
        the payment.

        sender_user_cellphone (str, optional): The celphone number of the
        user sending the payment.

        receiver_user_id (str): The id of the user receiving the payment.

        receiver_user_email (str): The email of the user receiving the payment.

        pay_object_id (str): A payment identifier in the TPA context.

        amount_BIT (float, optional): The amount of bitcoins to be payed by the
        widget. If not specified here, it must be entered on payment basis.

        pay_type (str): The string representing the type of operation
        ("Tip", "Pay", "Deposit" or "Donate").

        reference_code (str, optional): A custom code to be tracked by the
        TPA. It's sent back to the TPA in the specified callback. It could be
        used also to search with the micro payments query API.

        end_mpayment_uri (str, optional): The callback URL to notify a
        successful micro payment. The callback will be called with parameters
        "reference_code" and "request_UID".

        end_mpayment_redirect_uri (str, optional): An URL to be redirected to
        after a successful micro payment.

        redirect_uri (str, optional): redirect URL after a successful OAuth 
        flow. The URL must accept a "code" parameter if access is granted or
        "error" and "error_description" in case of denial.
    """
    def __init__(self, sender_user_id="", sender_user_email="",
                 sender_user_cellphone="", receiver_user_id="",
                 receiver_user_email="", pay_object_id="", amount_BIT=0,
                 timestamp=int(round(time.time() * 1000)),
                 pay_type=PaymentType.NONE, reference_code="", 
                 end_mpayment_uri="", end_mpayment_redirect_uri="",
                 redirect_uri=""):
        self.sender_user_id = sender_user_id
        self.sender_user_email = sender_user_email
        self.sender_user_cellphone = sender_user_cellphone
        self.receiver_user_id = receiver_user_id
        self.receiver_user_email = receiver_user_email
        self.pay_object_id = pay_object_id
        self.amount_BIT = amount_BIT
        self.timestamp = timestamp
        self.pay_type = pay_type
        self.reference_code = reference_code
        self.end_mpayment_uri = end_mpayment_uri
        self.end_mpayment_redirect_uri = end_mpayment_redirect_uri
        self.redirect_uri = redirect_uri


class MicroPaymentCustomization:
    """ Micro payment button configuration options.

    This function is intended to be a helper for creating empty micro
    payments buttons customization but also serves for documenting. A
    hash with the intended fields would give the same results.

    Attributes:
        login_cellphone_header_title+ (str, optional): Text to appear in the
        login screen. Default: "Support content creators by sending them bits.
        New users receive 50 bits to get started!"

        predefined_pay_values+ (str, optional): A string of comma separated
        amount values, e.g. "1,5,10".

        button_css+ (str, optional): optional CSS button customization
        ("red" | "grey").
    """
    def __init__(self, login_cellphone_header_title="",
                 predefined_pay_values="", button_css=""):
        self.login_cellphone_header_title = login_cellphone_header_title
        self.predefined_pay_values = predefined_pay_values
        self.button_css = button_css


class MicroPayment:
    """ Xapo's payment buttons snippet builder.

    This class allows the construction of 2 kind of widgets, *div* and
    *iframe*. The result is a HTML snippet that could be embedded in a
    web page for doing micro payments though a payment button.

    Attributes:
        service_url (str): The endpoint URL that returns the payment widget.

        app_id (str, optional): The id of the TPA for which the widget will
        be created.

        app_secret (str, optional): The TPA secret used to encrypt widget
        configuration.
    """

    def __init__(self, service_url, app_id=None, app_secret=None):
        self.service_url = service_url
        self.app_id = app_id
        self.app_secret = app_secret

    def __build_url(self, config, customization):
        json_config = json.dumps(config.__dict__)

        if (self.app_id is None or self.app_secret is None):
            query = {
                "payload": json_config,
                "customization": json.dumps(customization.__dict__)
            }
            query_str = urlencode(query)
        else:
            encrypted_config = xapo_utils.encrypt(json_config, self.app_secret,
                                                  xapo_utils.pkcs7_padding)
            query = {
                "app_id": self.app_id,
                "button_request": encrypted_config,
                "customization": json.dumps(customization.__dict__)
            }

        query_str = urlencode(query)
        widget_url = self.service_url + "?" + query_str

        return widget_url

    def build_iframe_widget(self, config, customization):
        """ Build an iframe HTML snippet in order to be embedded in apps.

        Args:
            config (MicroPaymentConfig): The button configuration options.
            See @MicroPaymentConfig.

        Returns:
            str: the iframe HTML snippet ot be embedded in a page.
        """
        widget_url = self.__build_url(config, customization)
        snippet = """
                <iframe id="tipButtonFrame" scrolling="no" frameborder="0"
                    style="border:none; overflow:hidden; height:22px;"
                    allowTransparency="true" src="{url}">
                </iframe>
              """.format(url=widget_url)

        return textwrap.dedent(snippet)

    def build_div_widget(self, config, customization):
        """ Build div HTML snippet in order to be embedded in apps.

        Args:
            config (MicroPaymentConfig): The button configuration options.
            See @MicroPaymentConfig.

        Returns:
            str: the div HTML snippet ot be embedded in a page.
        """
        widget_url = self.__build_url(config, customization)
        snippet = r"""
                <div id="tipButtonDiv" class="tipButtonDiv"></div>
                <div id="tipButtonPopup" class="tipButtonPopup"></div>
                <script>
                    $(document).ready(function() {{
                        $("#tipButtonDiv").load("{url}");
                    }});
                </script>
              """.format(url=widget_url)

        return textwrap.dedent(snippet)


xapo_sdk/xapo_utils.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
from __future__ import absolute_import, division, print_function, \
    with_statement, unicode_literals
import base64
from Crypto.Cipher import AES


def pkcs7_padding(bytestring, k=16):
    """ Pad an input bytestring according to PKCS#7.

    Args:
        bytestring (str): The text to encode.

        k (int, optional): The padding block size. It defaults to k=16.

    Returns:
        str: The padded bytestring.
    """
    l = len(bytestring)
    val = k - (l % k)

    return bytestring + bytearray([val] * val).decode()


def zero_padding(bytestring, k=16):
    l = len(bytestring)
    val = k - (l % k)

    return bytestring + bytearray([0] * val).decode()


def encrypt(payload, app_secret, padding=zero_padding):
    """ Payload encrypting function.

    Pad and encrypt the payload in order to call XAPO Bitcoin API.

    Args:
        payload (str): The payload to be encrypted.

        app_secret (str): The key used to encrypt the payload.

    Returns:
        str: A string with the encrypted and base64 encoded payload.
    """
    cipher = AES.new(key=app_secret, mode=AES.MODE_ECB)
    padded_payload = padding(payload)
    encrypted_payload = cipher.encrypt(padded_payload)
    encoded_payload = base64.b64encode(encrypted_payload)

    return encoded_payload


xapo_sdk/micro_payment.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from __future__ import absolute_import, division, print_function, \
    with_statement, unicode_literals

from future.standard_library import hooks
with hooks():
    from urllib.parse import urlencode

import time
import base64
import textwrap
import json
from Crypto.Cipher import AES


def _pkcs7_encode(bytestring, k=16):
    """
    Pad an input bytestring according to PKCS#7.

    Args:
        bytestring (str): The text to encode.
        k (int, optional): The padding block size. It defaults to k=16.

    Returns:
        str: The padded bytestring.
    """
    l = len(bytestring)
    val = k - (l % k)
    return bytestring + bytearray([val] * val).decode()


# TODO probably this should be moved to xapo_utils.py in order to be
#      reused when the SDK and tools grow up.
def _encrypt(payload, app_secret):
    """ Payload encrypting function.

    Pad and encrypt the payload in order to call XAPO Bitcoin API.

    Args:
        payload (str): The payload to be encrypted.
        app_secret (str): The key used to encrypt the payload.

    Returns:
        str: A string with the encrypted and base64 encoded payload.
    """
    cipher = AES.new(key=app_secret, mode=AES.MODE_ECB)
    padded_payload = _pkcs7_encode(payload)
    encrypted_payload = cipher.encrypt(padded_payload)
    encoded_payload = base64.b64encode(encrypted_payload)

    return encoded_payload


class MicroPaymentConfig:
    """ Micro payment button configuration options.

    This class is intended to be a placeholder for micro payments
    buttons configuration but also serves for documenting. A dictionary
    with the intended fields would give the same results.

    Attributes:
        sender_user_id (str): The id of the user sending the payment.
        sender_user_email (str, optional): The email of the user sending
            the payment.
        sender_user_cellphone (str, optional): The celphone number of the user
            sending the payment.
        receiver_user_id (str): The id of the user receiving the payment.
        receiver_user_email (str): The email of the user receiving the payment.
        pay_object_id (str): A payment identifier in the TPA context.
        amount_BIT (float, optional): The amount of bitcoins to be payed by the
            widget. If not specified here, it must be entered on payment basis.
    """
    def __init__(self, sender_user_id="", sender_user_email="",
                 sender_user_cellphone="", receiver_user_id="",
                 receiver_user_email="", pay_object_id="", amount_BIT=0,
                 timestamp=int(round(time.time() * 1000))):
        self.sender_user_id = sender_user_id
        self.sender_user_email = sender_user_email
        self.sender_user_cellphone = sender_user_cellphone
        self.receiver_user_id = receiver_user_id
        self.receiver_user_email = receiver_user_email
        self.pay_object_id = pay_object_id
        self.amount_BIT = amount_BIT
        self.timestamp = timestamp


class XapoMicroPaymentSDK:
    """ Xapo's payment buttons snippet builder.

    This class allows the construction of 2 kind of widgets, *div* and
    *iframe*. The result is a HTML snippet that could be embedded in a
    web page for doing micro payments though a payment button.

    Attributes:
        service_url: The endpoint URL that returns the payment widget.
        app_id: The id of the TPA for which the widget will be created.
        app_secret: The TPA secret used to encrypt widget configuration.
    """

    def __init__(self, service_url, app_id, app_secret):
        self.service_url = service_url
        self.app_id = app_id
        self.app_secret = app_secret

    def __build_url(self, config, pay_type):
        json_config = json.dumps(config.__dict__)
        encrypted_config = _encrypt(json_config, self.app_secret)

        query = {"app_id": self.app_id, "button_request": encrypted_config,
                 "customization": json.dumps({"button_text": pay_type})}
        query_str = urlencode(query)

        widget_url = self.service_url + "?" + query_str

        return widget_url

    def build_iframe_widget(self, config, pay_type):
        """ Build an iframe HTML snippet in order to be embedded in apps.

        Args:
            config (MicroPaymentConfig): The button configuration options.
                See @MicroPaymentConfig.
            pay_type (str): The string representing the type of operation
                      ("Tip", "Pay", "Deposit" or "Donate").

        Returns:
            string: the iframe HTML snippet ot be embedded in a page.

        Example:
        >>> xmp = XapoMicroPaymentSDK(
        ... "http://dev.xapo.com:8089/pay_button/show",
        ... "b91014cc28c94841",
        ... "c533a6e606fb62ccb13e8baf8a95cbdc")
        >>> mpc = MicroPaymentConfig(
        ... sender_user_email="sender@xapo.com",
        ... sender_user_cellphone="+5491112341234",
        ... receiver_user_id="r0210",
        ... receiver_user_email="fernando.taboada@xapo.com",
        ... pay_object_id="to0210",
        ... amount_BIT=0.01)
        >>> iframe = xmp.build_iframe_widget(mpc, pay_type = "Tip")
        >>> print(iframe) # doctest: +ELLIPSIS
        <BLANKLINE>
        <iframe...</iframe>
        <BLANKLINE>
        """
        # TODO see if pay_type should be handled like this, in PHP it's part
        #      of the "request". In this implementation we decoupled the
        #      request from the config object for grater flexibility
        widget_url = self.__build_url(config, pay_type)
        snippet = """
                <iframe id="tipButtonFrame" scrolling="no" frameborder="0"
                    style="border:none; overflow:hidden; height:22px;"
                    allowTransparency="true" src="{url}">
                </iframe>
              """.format(url=widget_url)

        return textwrap.dedent(snippet)

    def build_div_widget(self, config, pay_type):
        """ Build div HTML snippet in order to be embedded in apps.

        Args:
            config (MicroPaymentConfig): The button configuration options.
                See @MicroPaymentConfig.
            pay_type (str): The string representing the type of operation
                      ("Tip", "Pay", "Deposit" or "Donate").

        Returns:
            string: the div HTML snippet ot be embedded in a page.

        Example:
        >>> xmp = XapoMicroPaymentSDK(
        ... "http://dev.xapo.com:8089/pay_button/show",
        ... "b91014cc28c94841",
        ... "c533a6e606fb62ccb13e8baf8a95cbdc")
        >>> mpc = MicroPaymentConfig(
        ... sender_user_email="sender@xapo.com",
        ... sender_user_cellphone="+5491112341234",
        ... receiver_user_id="r0210",
        ... receiver_user_email="fernando.taboada@xapo.com",
        ... pay_object_id="to0210",
        ... amount_BIT=0.01)
        >>> div = xmp.build_div_widget(mpc, pay_type = "Tip")
        >>> print(div) # doctest: +ELLIPSIS
        <BLANKLINE>
        <div id="tipButtonDiv" class="tipButtonDiv"></div>
        <div id="tipButtonPopup" class="tipButtonPopup"></div>
        <script>...</script>
        <BLANKLINE>
        """
        widget_url = self.__build_url(config, pay_type)
        snippet = r"""
                <div id="tipButtonDiv" class="tipButtonDiv"></div>
                <div id="tipButtonPopup" class="tipButtonPopup"></div>
                <script>
                    $(document).ready(function() {{
                        $("#tipButtonDiv").load("{url}");
                    }});
                </script>
              """.format(url=widget_url)

        return textwrap.dedent(snippet)

if __name__ == '__main__':
    import doctest
    doctest.testmod(verbose=True)
